compiler/rustc_mir_build/src/thir/pattern/deconstruct_pat.rs

```diff
@@ -1186,17 +1245,16 @@ impl ConstructorSet {
                 present.extend(seen);
                 missing.push(NonExhaustive);
             }
-            // If `exhaustive_patterns` is disabled and our scrutinee is an empty type, we cannot
-            // expose its emptiness. The exception is if the pattern is at the top level, because we
-            // want empty matches to be considered exhaustive.
-            ConstructorSet::Uninhabited
-                if !pcx.cx.tcx.features().exhaustive_patterns && !pcx.is_top_level =>
-            {
-                missing.push(NonExhaustive);
+            ConstructorSet::NoConstructors => {
+                if !pcx.is_top_level {
+                    missing_empty.push(NonExhaustive);
+                }
             }
-            ConstructorSet::Uninhabited => {}
         }
 
+        if !pcx.cx.tcx.features().exhaustive_patterns {
+            missing.extend(missing_empty);
+        }
```

Comment:
So this is why we don't need to check exhaustive_patterns in `is_uninhabited`? 
Because we now *always* characterize the ctor as missing-but-empty, and just treat them as missing here at the end?

Reply:
Yep exactly. Before this PR we kinda lied by pretending constructors were nonempty when `exhaustive_patterns` is off. Now we properly track which constructor is or isn't empty in `ConstructorSet`, and decide whether to count them as missing here